### PR TITLE
Boilerplate change- Switch from using opacity and settimeout to using visibility and keyframe animation.

### PR DESCRIPTION
**Pull request by sriramkrish85**
Switch from using opacity and settimeout
to using visibility and keyframe animation. #323 
Detailed proposal can be viewed here:
https://docs.google.com/document/d/1gZFaKvcDffceJNaI3bYfuYPtYU5u2y6UhE5wBPTsJ9w

### CODE REVIEW

File: src/styles.js [hunk 1]
```diff
@@ -81,8 +82,11 @@ export function installStyles(doc, cssText, cb, opt_isRuntimeCss) {
 export function makeBodyVisible(doc) {
   let interval;
   const set = () => {
+
     if (doc.body) {
       doc.body.style.opacity = 1;
```
Comment (dvoytenko) [inline src/styles.js]:
```
setStyle(doc.body, {
  opacity: 1,
  visibility: 'visible',
  animation: 'none'
})
```

File: test/fixtures/boilerplate-visibility.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<!doctype html>
+<html ⚡>
+<head>
+  <meta charset="utf-8">
+  <title>Released AMP components</title>
+  <link rel="canonical" href="amps.html" >
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+
+  <style>body{-webkit-animation:-amp-start 1s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 1s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 1s steps(1,end) 0s 1 normal both;animation:-amp-start 1s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
```
Comment (dvoytenko) [inline test/fixtures/boilerplate-visibility.html]: One note to make here. Ideally I'd like to tag this style with some attribute, e.g. `<style amp-starter>` or such. I'd prefer this to our current system where we do the opposite and use `<style amp-custom>` for the user-defined styles. Boilerplate is intended to be copy-pasted into HTML and thus extra few chars don't hurt. Then, of course, I'd like not to have it at all. Please include as a separate section in our doc and call it out directly for the validator bug.

File: test/functional/test-styles.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  it('makeBodyVisible', () => {
+    styles.makeBodyVisible(document);
+    clock.tick(1000);
+    expect(document.body).to.exist;
+    expect(document.body.style.opacity).to.equal('1');
+    expect(getStyle(document.body, 'visibility')).to.equal('visible');
+    expect(getStyle(document.body, 'animation')).to.equal('none');
+  });
+
```
Comment (dvoytenko) [inline test/functional/test-styles.js]: nit: remove extra lines

File: test/integration/test-boilerplates.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createFixtureIframe, expectBodyToBecomeVisible} from
+    '../../testing/iframe.js';
+import {getStyle} from '../../src/style';
+
+describe('Opacity Boilerplate', () => {
```
Comment (dvoytenko) [inline test/integration/test-boilerplates.js]: Let's call it "OLD Opacity Booilerplate"

File: test/integration/test-boilerplates.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  it('should show the body when opacity boilerplate is used', () => {
+    return expectBodyToBecomeVisible(fixture.win).then(() => {
+      expect(getStyle(fixture.win.document.body, 'opacity')).to.equal('1');
+    });
+  });
+});
+
+
+describe('Visibility Boilerplate', () => {
```
Comment (dvoytenko) [inline test/integration/test-boilerplates.js]: By extension "NEW" or "V0" and "V1"

File: test/integration/test-boilerplates.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import {createFixtureIframe, expectBodyToBecomeVisible} from
+    '../../testing/iframe.js';
+import {getStyle} from '../../src/style';
+
+describe('Opacity Boilerplate', () => {
+
+  let fixture;
+  beforeEach(() => {
+    return createFixtureIframe('test/fixtures/boilerplate-opacity.html', 1000)
```
Comment (dvoytenko) [inline test/integration/test-boilerplates.js]: Same with file names: let's make sure the version of the bolierplate is clear. OLD/NEW or V0/V1.

**Comment by dvoytenko** [comment]
All looks great. Just few minor comments. One thing, though, please make the research/spec doc public and link it from the description of this PR.

File: src/styles.js (hunk 1 above)
Comment (sriramkrish85) [inline src/styles.js]: Done.

File: test/functional/test-styles.js (hunk 3 above)
Comment (sriramkrish85) [inline test/functional/test-styles.js]: Done.

File: test/integration/test-boilerplates.js (hunk 4 above)
Comment (sriramkrish85) [inline test/integration/test-boilerplates.js]: Done.

File: test/integration/test-boilerplates.js (hunk 5 above)
Comment (sriramkrish85) [inline test/integration/test-boilerplates.js]: Done.

File: test/integration/test-boilerplates.js (hunk 6 above)
Comment (sriramkrish85) [inline test/integration/test-boilerplates.js]: Done.

File: src/styles.js (hunk 1 above)
Comment (erwinmombay) [inline src/styles.js]: dont we want `setsStyles`

File: src/styles.js [hunk 7]
```diff
@@ -82,7 +83,11 @@ export function makeBodyVisible(doc) {
   let interval;
   const set = () => {
     if (doc.body) {
-      doc.body.style.opacity = 1;
+      setStyle(doc.body, {
```
Comment (dvoytenko) [inline src/styles.js]: It should be setStyles (notice 's')

**Comment by dvoytenko** [comment]
LGTM once setStyles and tests are fixed.

**Comment by jmadler** [comment]
/cc @Meggin, @pbakaus 

Will the validator be updated as well?

**Comment by amphtml-team** [comment]
Yes. We won't make the change public until the validator is deployed.

**Comment by jmadler** [comment]
You'll probably need to update the samples as well, so that they'll validate

**Comment by amphtml-team** [comment]
Will do

**Comment by dvoytenko** [comment]
@sriramkrish85 This is ready to be merged. Right?

**Comment by amphtml-team** [comment]
Yes ,will have to update example files once validator is ready. Else they
will error out !

**Comment by dvoytenko** [comment]
Let's merge this PR. We will update examples and documentation once validator is ready.

**Comment by adactio** [comment]
I'm not sure this issue should be closed as long as the documentation is still saying that the style with opacity:0 is required:

https://github.com/ampproject/amphtml
https://www.ampproject.org/docs/get_started/create/basic_markup.html

Or should another issue be opened for updating the docs?

**Comment by jridgewell** [comment]
@adactio: We still have https://github.com/ampproject/amphtml/issues/323 open until the doc updates are in.

**Pull request merged**